File: scripts/src/main/kotlin/scripts/updateAuthors.kt
/*
 * Copyright 2003-2025 The IdeaVim authors
 *
 * Use of this source code is governed by an MIT-style
 * license that can be found in the LICENSE.txt file or at
 * https://opensource.org/licenses/MIT.
 */

package scripts

import org.eclipse.jgit.api.Git
import org.eclipse.jgit.lib.RepositoryBuilder
import org.kohsuke.github.GHUser
import java.io.File

data class Author(val name: String, val url: String, val mail: String) {
  val isJetBrainsEmployee: Boolean
    get() = mail.endsWith("@jetbrains.com")
}

private const val CONTRIBUTORS_HEADER = "Contributors:"
private const val JETBRAINS_IP_HEADER = "Contributors with JetBrains IP:"
private const val PREVIOUS_CONTRIBUTORS_HEADER = "Previous contributors:"

fun main(args: Array<String>) {
  val projectDir = if (args.isNotEmpty()) File(args[0]) else File(".")

  val uncheckedEmails = setOf(
    "aleksei.plate@jetbrains.com",
    "aleksei.plate@teamcity",
    "aleksei.plate@TeamCity",
    "alex.plate@192.168.0.109",
    "nikita.koshcheev@TeamCity",
    "TeamCity@TeamCity",
  )

  updateAuthors(projectDir, uncheckedEmails)
}

fun updateAuthors(projectDir: File, uncheckedEmails: Set<String>) {
  println("Start update authors")
  println(projectDir)
  val repository = RepositoryBuilder().setGitDir(File(projectDir, ".git")).build()
  val git = Git(repository)
  val lastSuccessfulCommit = System.getenv("SUCCESS_COMMIT")!!
  val hashesAndEmails = git.log().call()
    .takeWhile {
      !it.id.name.equals(lastSuccessfulCommit, ignoreCase = true)
    }
    .associate { it.authorIdent.emailAddress to it.name }

  println("Last successful commit: $lastSuccessfulCommit")
  println("Amount of commits: ${hashesAndEmails.size}")
  println("Emails: ${hashesAndEmails.keys}")
  val gitHub = org.kohsuke.github.GitHub.connect()
  val ghRepository = gitHub.getRepository("JetBrains/ideavim")
  val users = mutableSetOf<Author>()
  println("Start emails processing")
  for ((email, hash) in hashesAndEmails) {
    println("Processing '$email'...")
    if (email in uncheckedEmails) {
      println("Email '$email' is in unchecked emails. Skip it")
      continue
    }
    if ("[bot]@users.noreply.github.com" in email) {
      println("Email '$email' is from a bot. Skip it")
      continue
    }
    if ("tcuser" in email) {
      println("Email '$email' is from teamcity. Skip it")
      continue
    }
    val user: GHUser? = ghRepository.getCommit(hash).author
    if (user == null) {
      println("Cant get the commit author. Email: $email. Commit: $hash")
      continue
    }
    val htmlUrl = user.htmlUrl.toString()
    val name = user.name ?: user.login
    users.add(Author(name, htmlUrl, email))
  }

  println("Emails processed")
  val authorsFile = File(projectDir, "AUTHORS.md")
  val authorsContent = authorsFile.readText()

  val result = addAuthorsToContent(authorsContent, users)
  if (result.newAuthors.isNotEmpty()) {
    val authorNames = result.newAuthors.joinToString(", ") { it.name }
    println("::set-output name=authors::$authorNames")
    authorsFile.writeText(result.content)
  }
}

data class AddAuthorsResult(
  val content: String,
  val newAuthors: List<Author>,
)

/**
 * Adds authors to the AUTHORS.md content.
 * - JetBrains employees (emails ending with @jetbrains.com) are added to "Contributors with JetBrains IP:" section
 * - Other contributors are added to "Contributors:" section
 * - Authors already present in either section (by email or GitHub URL) are skipped
 */
fun addAuthorsToContent(authorsContent: String, authors: Collection<Author>): AddAuthorsResult {
  val existingEmails = extractExistingEmails(authorsContent)
  val existingGitHubUrls = extractExistingGitHubUrls(authorsContent)

  val newAuthors = authors.filterNot { it.mail in existingEmails || it.url in existingGitHubUrls }
  if (newAuthors.isEmpty()) {
    return AddAuthorsResult(authorsContent, emptyList())
  }

  val jetBrainsAuthors = newAuthors.filter { it.isJetBrainsEmployee }
  val regularAuthors = newAuthors.filterNot { it.isJetBrainsEmployee }

  var result = authorsContent

  // Add JetBrains employees to JetBrains IP section
  if (jetBrainsAuthors.isNotEmpty()) {
    val insertionPoint = findSectionEndOffset(result, JETBRAINS_IP_HEADER, PREVIOUS_CONTRIBUTORS_HEADER)
    if (insertionPoint != -1) {
      val insertionString = jetBrainsAuthors.toMdString(isJetBrainsSection = true)
      result = StringBuilder(result).insert(insertionPoint, insertionString).toString()
    }
  }

  // Add regular contributors to Contributors section
  if (regularAuthors.isNotEmpty()) {
    val insertionPoint = findSectionEndOffset(result, CONTRIBUTORS_HEADER, JETBRAINS_IP_HEADER)
    if (insertionPoint != -1) {
      val insertionString = regularAuthors.toMdString(isJetBrainsSection = false)
      result = StringBuilder(result).insert(insertionPoint, insertionString).toString()
    }
  }

  return AddAuthorsResult(result, newAuthors)
}

/**
 * Extracts all email addresses from the AUTHORS.md content.
 * Emails are found in mailto: links.
 */
fun extractExistingEmails(content: String): Set<String> {
  return Regex("""mailto:([^)]+)\)""")
    .findAll(content)
    .map { it.groupValues[1] }
    .toSet()
}

/**
 * Extracts all GitHub profile URLs from the AUTHORS.md content.
 */
fun extractExistingGitHubUrls(content: String): Set<String> {
  return Regex("""\[!\[icon]\[github]]\((https://github\.com/[^)]+)\)""")
    .findAll(content)
    .map { it.groupValues[1] }
    .toSet()
}

/**
 * Finds the end offset of a section (where new entries should be inserted).
 * The section starts with [sectionHeader] and ends before [nextSectionHeader].
 * Returns the position just before the blank line preceding the next section header.
 */
fun findSectionEndOffset(content: String, sectionHeader: String, nextSectionHeader: String): Int {
  val sectionStart = content.indexOf(sectionHeader)
  if (sectionStart == -1) return -1

  val nextSectionStart = content.indexOf(nextSectionHeader, sectionStart)
  if (nextSectionStart == -1) return -1

  // Find the last non-blank line before the next section
  // We want to insert before the blank lines that precede the next section header
  var insertionPoint = nextSectionStart
  while (insertionPoint > sectionStart && content[insertionPoint - 1] == '\n') {
    insertionPoint--
  }

  return insertionPoint
}

fun List<Author>.toMdString(isJetBrainsSection: Boolean): String {
  return this.joinToString(separator = "") { author ->
    val nameWithNote = if (isJetBrainsSection) "${author.name} (JetBrains employee)" else author.name
    """
          |
          |* [![icon][mail]](mailto:${author.mail})
          |  [![icon][github]](${author.url})
          |  &nbsp;
          |  $nameWithNote
        """.trimMargin()
  }
}

File: scripts-ts/src/updateAuthors.ts
#!/usr/bin/env tsx
/**
 * Updates AUTHORS.md with new contributors from recent commits.
 *
 * Iterates commits since SUCCESS_COMMIT, looks up author info via GitHub API,
 * and adds new contributors to the appropriate section in AUTHORS.md.
 *
 * Environment:
 * - SUCCESS_COMMIT: Git hash of last successful workflow run
 * - GITHUB_TOKEN: GitHub token for API access
 */

import { execSync } from "child_process";
import { readFileSync, writeFileSync, appendFileSync } from "fs";
import { join } from "path";

export interface Author {
  name: string;
  url: string;
  mail: string;
}

export function isJetBrainsEmployee(author: Author): boolean {
  return author.mail.endsWith("@jetbrains.com");
}

const CONTRIBUTORS_HEADER = "Contributors:";
const JETBRAINS_IP_HEADER = "Contributors with JetBrains IP:";
const PREVIOUS_CONTRIBUTORS_HEADER = "Previous contributors:";

const uncheckedEmails = new Set([
  "aleksei.plate@jetbrains.com",
  "aleksei.plate@teamcity",
  "aleksei.plate@TeamCity",
  "alex.plate@192.168.0.109",
  "nikita.koshcheev@TeamCity",
  "TeamCity@TeamCity",
]);

function getCommitEmailsAndHashes(projectDir: string): Map<string, string> {
  const lastSuccessfulCommit = process.env.SUCCESS_COMMIT;
  if (!lastSuccessfulCommit) {
    throw new Error("SUCCESS_COMMIT environment variable is not set");
  }

  console.log(`Last successful commit: ${lastSuccessfulCommit}`);

  // Get commits with email and hash
  const gitOutput = execSync(
    `git log --format="%ae %H" ${lastSuccessfulCommit}..HEAD`,
    { cwd: projectDir, encoding: "utf-8" }
  );

  const result = new Map<string, string>();
  const lines = gitOutput.trim().split("\n").filter(Boolean);

  for (const line of lines) {
    const spaceIndex = line.indexOf(" ");
    if (spaceIndex !== -1) {
      const email = line.substring(0, spaceIndex);
      const hash = line.substring(spaceIndex + 1);
      if (!result.has(email)) {
        result.set(email, hash);
      }
    }
  }

  console.log(`Amount of commits: ${lines.length}`);
  console.log(`Unique emails: ${Array.from(result.keys()).join(", ")}`);

  return result;
}

async function getGitHubAuthorInfo(
  hash: string
): Promise<{ name: string; url: string } | null> {
  const token = process.env.GITHUB_TOKEN;
  if (!token) {
    throw new Error("GITHUB_TOKEN environment variable is not set");
  }

  try {
    const response = await fetch(
      `https://api.github.com/repos/JetBrains/ideavim/commits/${hash}`,
      {
        headers: {
          Authorization: `Bearer ${token}`,
          Accept: "application/vnd.github.v3+json",
        },
      }
    );

    if (!response.ok) {
      console.log(`Failed to fetch commit ${hash}: ${response.status}`);
      return null;
    }

    const data = await response.json();
    const author = data.author;

    if (!author) {
      return null;
    }

    return {
      name: author.name || author.login,
      url: author.html_url,
    };
  } catch (error) {
    console.log(`Error fetching commit ${hash}: ${error}`);
    return null;
  }
}

export function extractExistingEmails(content: string): Set<string> {
  const regex = /mailto:([^)]+)\)/g;
  const emails = new Set<string>();
  let match;
  while ((match = regex.exec(content)) !== null) {
    emails.add(match[1]);
  }
  return emails;
}

export function extractExistingGitHubUrls(content: string): Set<string> {
  const regex = /\[!\[icon]\[github]]\((https:\/\/github\.com\/[^)]+)\)/g;
  const urls = new Set<string>();
  let match;
  while ((match = regex.exec(content)) !== null) {
    urls.add(match[1]);
  }
  return urls;
}

export function findSectionEndOffset(
  content: string,
  sectionHeader: string,
  nextSectionHeader: string
): number {
  const sectionStart = content.indexOf(sectionHeader);
  if (sectionStart === -1) return -1;

  const nextSectionStart = content.indexOf(nextSectionHeader, sectionStart);
  if (nextSectionStart === -1) return -1;

  // Find the last non-blank line before the next section
  let insertionPoint = nextSectionStart;
  while (insertionPoint > sectionStart && content[insertionPoint - 1] === "\n") {
    insertionPoint--;
  }

  return insertionPoint;
}

function authorsToMdString(authors: Author[], isJetBrainsSection: boolean): string {
  return authors
    .map((author) => {
      const nameWithNote = isJetBrainsSection
        ? `${author.name} (JetBrains employee)`
        : author.name;
      return `
* [![icon][mail]](mailto:${author.mail})
  [![icon][github]](${author.url})
  &nbsp;
  ${nameWithNote}`;
    })
    .join("");
}

export interface AddAuthorsResult {
  content: string;
  newAuthors: Author[];
}

export function addAuthorsToContent(
  authorsContent: string,
  authors: Author[]
): AddAuthorsResult {
  const existingEmails = extractExistingEmails(authorsContent);
  const existingGitHubUrls = extractExistingGitHubUrls(authorsContent);

  const newAuthors = authors.filter(
    (a) => !existingEmails.has(a.mail) && !existingGitHubUrls.has(a.url)
  );

  if (newAuthors.length === 0) {
    return { content: authorsContent, newAuthors: [] };
  }

  const jetBrainsAuthors = newAuthors.filter(isJetBrainsEmployee);
  const regularAuthors = newAuthors.filter((a) => !isJetBrainsEmployee(a));

  let result = authorsContent;

  // Add JetBrains employees to JetBrains IP section
  if (jetBrainsAuthors.length > 0) {
    const insertionPoint = findSectionEndOffset(
      result,
      JETBRAINS_IP_HEADER,
      PREVIOUS_CONTRIBUTORS_HEADER
    );
    if (insertionPoint !== -1) {
      const insertionString = authorsToMdString(jetBrainsAuthors, true);
      result =
        result.slice(0, insertionPoint) +
        insertionString +
        result.slice(insertionPoint);
    }
  }

  // Add regular contributors to Contributors section
  if (regularAuthors.length > 0) {
    const insertionPoint = findSectionEndOffset(
      result,
      CONTRIBUTORS_HEADER,
      JETBRAINS_IP_HEADER
    );
    if (insertionPoint !== -1) {
      const insertionString = authorsToMdString(regularAuthors, false);
      result =
        result.slice(0, insertionPoint) +
        insertionString +
        result.slice(insertionPoint);
    }
  }

  return { content: result, newAuthors };
}

function writeGitHubOutput(name: string, value: string): void {
  const outputFile = process.env.GITHUB_OUTPUT;
  if (outputFile) {
    appendFileSync(outputFile, `${name}=${value}\n`);
  } else {
    console.log(`OUTPUT: ${name}=${value}`);
  }
}

async function main(): Promise<void> {
  const projectDir = process.argv[2] || ".";

  console.log("Start update authors");
  console.log(`Project directory: ${projectDir}`);

  const emailsAndHashes = getCommitEmailsAndHashes(projectDir);

  const users: Author[] = [];

  console.log("Start emails processing");
  for (const [email, hash] of emailsAndHashes) {
    console.log(`Processing '${email}'...`);

    if (uncheckedEmails.has(email)) {
      console.log(`Email '${email}' is in unchecked emails. Skip it`);
      continue;
    }

    if (email.includes("[bot]@users.noreply.github.com")) {
      console.log(`Email '${email}' is from a bot. Skip it`);
      continue;
    }

    if (email.includes("tcuser")) {
      console.log(`Email '${email}' is from teamcity. Skip it`);
      continue;
    }

    const authorInfo = await getGitHubAuthorInfo(hash);
    if (!authorInfo) {
      console.log(`Can't get the commit author. Email: ${email}. Commit: ${hash}`);
      continue;
    }

    users.push({
      name: authorInfo.name,
      url: authorInfo.url,
      mail: email,
    });
  }

  console.log("Emails processed");

  const authorsFilePath = join(projectDir, "AUTHORS.md");
  const authorsContent = readFileSync(authorsFilePath, "utf-8");

  const result = addAuthorsToContent(authorsContent, users);

  if (result.newAuthors.length > 0) {
    const authorNames = result.newAuthors.map((a) => a.name).join(", ");
    console.log(`New authors: ${authorNames}`);
    writeGitHubOutput("authors", authorNames);
    writeFileSync(authorsFilePath, result.content);
    console.log("AUTHORS.md updated");
  } else {
    console.log("No new authors to add");
  }
}

// Only run main when executed directly, not when imported
const isMainModule = import.meta.url === `file://${process.argv[1]}`;
if (isMainModule) {
  main().catch((error) => {
    console.error("Error:", error.message);
    process.exit(1);
  });
}
